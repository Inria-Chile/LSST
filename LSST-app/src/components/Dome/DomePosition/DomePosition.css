.dome-position-info {
    vertical-align: middle;
    display: inline-block;
    font-size: var(--dome-text-size);
}

.dome-angle {
    stroke: var(--theme-dome-angle);
    fill: var(--theme-dome-angle);
    stroke-width: 5;
    stroke-dasharray: 15;
}

.mount-angle {
    stroke-width: 3;
    stroke: var(--theme-dome-mount-angle);
}

.dome-azimuth-text {
    color: var(--theme-dome-angle);
}

.dome-data {
    color: var(--theme-active-text);
}