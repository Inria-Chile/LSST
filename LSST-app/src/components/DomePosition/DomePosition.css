.status-circle-open:before {
    content: ' \25CF';
    color: green
}

.status-circle-closed:before {
    content: ' \25CF';
    color: red
}

.dome-position-container {
    margin-top: 50px;
}

.dome-position-info {
    vertical-align: top;
    display: inline-block;
}