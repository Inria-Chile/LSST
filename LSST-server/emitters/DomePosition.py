import sys
import time
import datetime
import random

from SALPY_domeADB import *
from SALPY_domeLWS import *
from flask_socketio import send, emit
from flask_socketio import SocketIO

#Get data from ts_sal connection
def start_listening_dome_position(app, socketio):
    print("SAL starting")
    salAz = SAL_domeADB()
    salAz.setDebugLevel(0)

    salEl = SAL_domeLWS()
    salEl.setDebugLevel(0)

    topicDomeAz = domeADB_statusC()
    topicDomeEl = domeLWS_statusC()

    salAz.salTelemetrySub("domeADB_status")
    salEl.salTelemetrySub("domeLWS_status")

    print("SAL listening")
    try:
        while True:
            time.sleep(0.3)
            scodeAz = salAz.getSample_status(topicDomeAz)
            scodeEl = salEl.getSample_status(topicDomeEl)
            if scodeAz == 0 and scodeEl == 0:
                publish(app, socketio, topicDomeAz, topicDomeEl)

    except KeyboardInterrupt:
        print("SAL shutdown")
        salAz.salShutdown()
        salEl.salShutdown()
        sys.exit(0)
    
# Publish data to WS connection
def publish(app, socketio, topicDomeAz, topicDomeEl):
    print('Emitting DomePosition', [topicDomeAz.position_actual, topicDomeAz.position_error, topicDomeAz.position_cmd, topicDomeEl.position_actual, topicDomeEl.position_error, topicDomeEl.position_cmd])
    with app.test_request_context('/'):
        socketio.emit('DomePosition', {'DomeAzPos': topicDomeAz.position_actual, 'DomeAzPosErr':topicDomeAz.position_error, 'DomeAzCMD':topicDomeAz.position_cmd, 'DomeElPos': topicDomeEl.position_actual, 'DomeElPosErr':topicDomeEl.position_error, 'DomeElCMD':topicDomeEl.position_cmd}, namespace='/domeposition')
