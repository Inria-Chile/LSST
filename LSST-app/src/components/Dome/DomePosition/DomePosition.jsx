import React, { PureComponent } from 'react'
import './DomePosition.css';
import * as d3 from 'd3';
import {DEMO_MODE} from '../../Utils/Utils';

class DomePosition extends PureComponent {

    constructor(props){
        super(props);
        this.data = [];
        this.distanceToCamera = 12;
        this.cameraFOV = 20;
        this.state = {
            data: [],
        }

    }

    setDomeAzimuth = (angle) => {
        let duration = 500;
        if(DEMO_MODE)
            duration = 2000;
        let domeApertureAngle = 2*Math.atanh(this.props.shuttersAperture/(2*this.distanceToCamera));
        d3.select("#dome-angle").transition().duration(duration).attr("transform", "translate("+this.props.width/2+","+this.props.height/2+")rotate("+angle+")");
        d3.select("#dome-angle-top").transition().attr("y2", -this.props.height/2+this.props.height/2+this.props.width/2*Math.tan(domeApertureAngle/2));
        d3.select("#dome-angle-bot").transition().attr("y2", -this.props.height/2+this.props.height/2-this.props.width/2*Math.tan(domeApertureAngle/2));
    }

    setMountAzimuth = (angle) => {
        let duration = 500;
        if(DEMO_MODE)
            duration = 2000;
        d3.select("#mount-angle").transition().duration(duration).attr("transform", "translate("+this.props.width/2+","+this.props.height/2+")rotate("+angle+")");
        d3.select("#mount-angle-top").transition().attr("y2", -this.props.height/2+this.props.height/2+this.props.width/2*Math.tan(Math.PI/180*this.cameraFOV/2));
        d3.select("#mount-angle-bot").transition().attr("y2", -this.props.height/2+this.props.height/2-this.props.width/2*Math.tan(Math.PI/180*this.cameraFOV/2));
    }

    componentDidUpdate(prevProps, prevState){
        if(!document.hidden){
            window.requestAnimationFrame(
                () => {
                    this.setDomeAzimuth(this.props.domeAzimuth != null ? this.props.domeAzimuth : 0);
                    this.setMountAzimuth(this.props.mountAzimuth != null ? this.props.mountAzimuth : 0);
                }
            );
        }
    }

    componentDidMount() {
    }

    
    render() {
        return (
            <div className="dome-position-container" ref="container">
                {/* <h4>Dome Position</h4> */}
                {/* <div className="cardinal-points">
                </div> */}
                <div className="dome-wrapper">
                    <img className="north-img" src="/img/Cardinal_directions.svg" alt="North direction"></img>
                    <svg 
                        className="svg-container"
                        height={this.props.height+100}
                        width={this.props.width+100}>
                        <g id="dome-angle"
                            transform={"translate("+this.props.width/2+","+this.props.height/2+")"}
                            height={this.props.height}
                            width={this.props.width}>
                            <image id="dome-background" x={-this.props.width/2} y={-this.props.height/2} width={this.props.width} height={this.props.height} opacity={1.0} xlinkHref="/img/dome_top.png"/>
                            <line id="dome-angle-top"
                                className="dome-angle"
                                x1={0} y1={0} 
                                x2={-this.props.width/2} y2={0} />
                            <line id="dome-angle-bot"
                                className="dome-angle"
                                x1={0} y1={0} 
                                x2={-this.props.width/2} y2={0} />

                            <circle id={"test-circle"}
                                className="dome-angle"
                                cx={0} cy={0} r={5}
                                key={'overlay'}
                                pointerEvents="all"
                            />
                        </g>
                        <g id="mount-angle"
                            transform={"translate("+this.props.width/2+","+this.props.height/2+")"}
                            height={this.props.height}
                            width={this.props.width}>
                            <line id="mount-angle-top"
                                className="mount-angle"
                                x1={0} y1={0} 
                                x2={-this.props.width/2} y2={0} 
                                strokeWidth={4} />
                            <line id="mount-angle-bot"
                                className="mount-angle"
                                x1={0} y1={0} 
                                x2={-this.props.width/2} y2={0} 
                                strokeWidth={4} />
                        </g>
                    </svg>
                </div>                
                <div className="dome-position-info">
                    <div className="dome-azimuth-text">
                        <span>Dome azimuth: </span> 
                        <span>
                            { this.props.domeAzimuth  !== null ? this.props.domeAzimuth.toFixed(1)+'ยบ' : 'None'}
                        </span>
                    </div>
                    <div>
                        <span className="dome-data-label">Mount azimuth: </span> 
                        <span className="dome-data">
                            {this.props.mountAzimuth  !== null ? this.props.mountAzimuth.toFixed(1)+'ยบ' : 'None'}
                        </span>
                    </div>
                </div>
            </div>
        );
    }
}


export default DomePosition;