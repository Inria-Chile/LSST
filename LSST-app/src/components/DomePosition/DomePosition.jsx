import React, { Component } from 'react'
import './DomePosition.css';
import * as d3 from 'd3';

class DomePosition extends Component {

    constructor(props){
        super(props);
        this.data = [];
        this.distanceToCamera = 12;
        this.cameraFOV = 20;
        this.state = {
            data: [],
            cameraAngle: 0,
            domeAngle: 0
        }

    }

    setDomeAngle = (angle) => {
        let domeApertureAngle = 2*Math.atanh(this.props.aperture/(2*this.distanceToCamera));
        d3.select("#dome-angle").transition().duration(1000).attr("transform", "translate("+this.props.width/2+","+this.props.height/2+")rotate("+angle+")");
        d3.select("#dome-angle-top").transition().attr("y2", -this.props.height/2+this.props.height/2+this.props.width/2*Math.tan(domeApertureAngle/2));
        d3.select("#dome-angle-bot").transition().attr("y2", -this.props.height/2+this.props.height/2-this.props.width/2*Math.tan(domeApertureAngle/2));
    }

    setCameraAngle = (angle) => {
        
        d3.select("#camera-angle").transition().duration(1000).attr("transform", "translate("+this.props.width/2+","+this.props.height/2+")rotate("+angle+")");
        d3.select("#camera-angle-top").transition().attr("y2", -this.props.height/2+this.props.height/2+this.props.width/2*Math.tan(Math.PI/180*this.cameraFOV/2));
        d3.select("#camera-angle-bot").transition().attr("y2", -this.props.height/2+this.props.height/2-this.props.width/2*Math.tan(Math.PI/180*this.cameraFOV/2));

        
    }

    componentDidUpdate(prevProps, prevState){
        this.setDomeAngle(this.state.domeAngle);
            this.setCameraAngle(this.state.cameraAngle);
    }

    componentDidMount() {
        setInterval( () => {
            // let newAngle = Math.max(0, Math.ceil(Math.random()*360));
            let newAngle = this.state.domeAngle + ((Math.random()-0.5)*90);
            let newCameraAngle = newAngle + ((Math.random()-0.5)*30);
            this.setState({
                domeAngle: newAngle,
                cameraAngle: newCameraAngle
                // aperture: 11
            })
        }, 2000)
    }

    
    render() {
        return (
            <div className="dome-position-container" ref="container">
                <h4>Dome Position</h4>
                <svg 
                    className="svg-container"
                    height={this.props.height}
                    width={this.props.width}>
                    <g id="dome-angle"
                        transform={"translate("+this.props.width/2+","+this.props.height/2+")"}
                        height={this.props.height}
                        width={this.props.width}>
                        <image id="dome-background" x={-this.props.width/2} y={-this.props.height/2} width={this.props.width} height={this.props.height} xlinkHref="/img/dome_top.png"/>
                        <line id="dome-angle-top"
                            x1={0} y1={0} 
                            x2={-this.props.width/2} y2={0} 
                            strokeWidth={4} stroke={'#00aa00'} />
                        <line id="dome-angle-bot"
                            x1={0} y1={0} 
                            x2={-this.props.width/2} y2={0} 
                            strokeWidth={4} stroke={'#00aa00'} />

                        <circle id={"test-circle"}
                            cx={0}
                            cy={0}
                            key={'overlay'}
                            fill={'#00aa00'}
                            r={5}
                            pointerEvents="all"
                        />
                    </g>
                    <g id="camera-angle"
                        transform={"translate("+this.props.width/2+","+this.props.height/2+")"}
                        height={this.props.height}
                        width={this.props.width}>
                        <line id="camera-angle-top"
                            x1={0} y1={0} 
                            x2={-this.props.width/2} y2={0} 
                            strokeWidth={4} stroke={'#ee0000'} />
                        <line id="camera-angle-bot"
                            x1={0} y1={0} 
                            x2={-this.props.width/2} y2={0} 
                            strokeWidth={4} stroke={'#ee0000'} />
                    </g>
                </svg>
                <div className="dome-position-info">
                    Dome azimuth angle: {this.state.domeAngle.toFixed(1)}ยบ <br/>
                    Camera azimuth angle: {this.state.cameraAngle.toFixed(1)}ยบ <br/>
                    Camera FOV: {this.props.aperture}ยบ <br/>
                </div>
            </div>
        );
    }
}


export default DomePosition;