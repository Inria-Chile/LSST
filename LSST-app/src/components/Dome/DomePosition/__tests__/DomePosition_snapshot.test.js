import React from 'react';
import DomePosition from '../DomePosition';
import renderer from 'react-test-renderer';


/*---- Snapshot testing ----*/
describe('DomePosition test', function(){
    const shuttersAperture = 10 ;
    const domeAzimuth = 10;
    const telescopeElevation = 10;
    const mountAzimuth = 10;
    const updateDomePos = 10;
    it('renders correctly',()=>{
        const tree = renderer
            .create(<DomePosition width={330} height={300} id="dome-top" scale={1.4} xOffset={-0.05} yOffset={0.15}
            shuttersAperture={shuttersAperture} domeAzimuth={domeAzimuth}
            telescopeElevation={telescopeElevation} mountAzimuth={mountAzimuth} updateDomePos={updateDomePos}/>) 
            .toJSON();
        expect(tree).toMatchSnapshot();
    })
})

