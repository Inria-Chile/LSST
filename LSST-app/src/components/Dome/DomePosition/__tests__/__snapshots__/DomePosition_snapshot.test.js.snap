// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomePosition test renders correctly 1`] = `
<div
  className="dome-position-container"
>
  <div
    className="cardinal-points"
  >
    <img
      alt="cardinal-points"
      src="/img/Cardinal_directions.png"
    />
  </div>
  <svg
    className="svg-container"
    height={400}
    width={430}
  >
    <g
      height={300}
      id="dome-angle"
      transform="translate(165,150)"
      width={330}
    >
      <image
        height={300}
        id="dome-background"
        opacity={1}
        width={330}
        x={-165}
        xlinkHref="/img/dome_top.png"
        y={-150}
      />
      <line
        className="dome-angle"
        id="dome-angle-top"
        x1={0}
        x2={-165}
        y1={0}
        y2={0}
      />
      <line
        className="dome-angle"
        id="dome-angle-bot"
        x1={0}
        x2={-165}
        y1={0}
        y2={0}
      />
      <circle
        className="dome-angle"
        cx={0}
        cy={0}
        id="test-circle"
        pointerEvents="all"
        r={5}
      />
    </g>
    <g
      height={300}
      id="mount-angle"
      transform="translate(165,150)"
      width={330}
    >
      <line
        className="mount-angle"
        id="mount-angle-top"
        strokeWidth={4}
        x1={0}
        x2={-165}
        y1={0}
        y2={0}
      />
      <line
        className="mount-angle"
        id="mount-angle-bot"
        strokeWidth={4}
        x1={0}
        x2={-165}
        y1={0}
        y2={0}
      />
    </g>
  </svg>
  <div
    className="dome-position-info"
  >
    <div
      className="dome-azimuth-text"
    >
      <span>
        Dome azimuth: 
      </span>
      <span>
        10.0ยบ
      </span>
    </div>
    <div>
      <span
        className="dome-data-label"
      >
        Mount azimuth: 
      </span>
      <span
        className="dome-data"
      >
        10.0ยบ
      </span>
    </div>
    <div>
      <span
        className="dome-data-label"
      >
        Camera FOV: 
      </span>
      <span
        className="dome-data"
      >
        10ยบ
      </span>
    </div>
  </div>
</div>
`;
